### PR TITLE
`opam monorepo` support

### PR DESCRIPTION
**Pull request by RyanGibb**
This PR adds support for [opam-monorepo](https://github.com/tarides/opam-monorepo).

To support building [mirage](https://mirage.io/) unikernels for cross-compilation targets we need to build them with dune since [mirage 4](https://mirage.io/docs/mirage-4) (in the absence of ocaml/opam cross compilation support).

Dune supports cross compilation with 'build contexts' using a toolchain which will use a target compiler installed in `<opam-switch>/default/<toolchain>` (as opposed to the host compiler). But not all dependencies are build with the target compiler. Notably ppx requires building with the host compiler. This information in stored in dune files -- e.g. [here](https://github.com/mirage/mirage-tcpip/blob/3ab30ab7b43dede75abf7b37838e051e0ddbb23a/src/tcp/dune#L9). In order to statically know how to build dependencies, we would have to parse the dune files to try and find out (which is non-trivial). Potentially a better solution is to encode the dependency build contexts in `opam` files, but opam [doesn't currently support cross-compilation](https://github.com/ocaml/opam/issues/1536). 

Instead, `opam-monorep` is used to build mirage unikernels. This fetches all the dependencies (filtered with `?monorepo` in the opam file)  to a local directory `duniverse`. Dune then builds the whole project. Unfortunately we can't benefit from the dune cache in our nix derivation, so all the dependencies will be rebuilt on every nix build.

This is implemented in the [opam-monorepo](https://github.com/tarides/opam-monorepo) project. This PR adds support for this workflow in the for of the `queryToScopeMonorepo` function, along with the lower layer functions `mkSrcsScope`, `deduplicateSrcs`, and `defsToSrcs`. We refactor the source fetching logic into `evaluator/default.nix` to share it between `mkSrc` (a derivation fetching a package's source) and `builder`.

To see an example of this usage see [https://github.com/RyanGibb/mirage-hello](https://github.com/RyanGibb/mirage-hello). A flake template is the works.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -582,6 +582,31 @@ Then, import it:
...
+ResolveArgs :
+{ env = ?{ ${var_name} = value : String; ... }
+; with-test = ?Bool
+; with-doc = ?Bool
+; dev = ?Bool
+; depopts = ?Bool
+; best-effort = ?Bool
+}
+```
+
```
Comment (balsoft) [inline README.md]: Isn't this already defined elsewhere?

File: src/opam.nix [hunk 2]
```diff
@@ -382,6 +382,77 @@ in rec {
...
+            cp -R . $out
+          '';
+        });
+    in
+    makeScope callPackageWith (self: builtins.listToAttrs (builtins.map
+      (src: lib.attrsets.nameValuePair src.name (mkSrc src)) srcs));
+
+  queryToScopeMonorepo = { repos ? [ opamRepository ], overlays ? [ ], resolveArgs ? {
+        # pass monorepo = 1 to `opam admin list` to pick up dependencies marked with {?monorepo}
+        env.monorepo=1;
```
Comment (balsoft) [inline src/opam.nix]: Why not move this down to the `opamList` call itself, so that providing a `resolveArgs` argument doesn't override this setting (which is, AFAIU, always useful for `queryToScopeMonorepo`)?

File: src/opam.nix [hunk 3]
```diff
@@ -382,6 +382,77 @@ in rec {
...
+      # > a single one. Here we decided to go with the one associated with the package
+      # > that has the higher version. We need a better long term solution as this won't
+      # > play nicely with pins for instance.
+      # > The best solution here would be to use source trimming, so we can pull each individual
+      # > package to its own directory and strip out all the unrelated source code but we would
+      # > need dune to provide that feature.
+      # See [opam-monorepo](https://github.com/tarides/opam-monorepo/blob/9262e7f71d749520b7e046fbd90a4732a43866e9/lib/duniverse.ml#L143-L157)
+      else if duplicateSrc.version >= newSrc.version then srcs
+      else (lib.lists.remove duplicateSrc srcs) ++ [ newSrc ];
+    in builtins.foldl' op [ ] srcs;
```
Comment (balsoft) [inline src/opam.nix]: Appreciate the comments here, really useful!

File: src/opam.nix [hunk 4]
```diff
@@ -382,6 +382,77 @@ in rec {
...
+  
+  deduplicateSrcs = srcs:
+    # This is O(n^2). We could try and improve this by sorting the list on name. But n is small.
+    let op = srcs: newSrc:
+      # Find if two packages come from the same dev-repo.
+      # Note we are assuming no dev-repos will have different names here, but we also assume
+      # this later when we will symlink in the duniverse directory based on this name.
+      let duplicateSrc = lib.lists.findSingle (src: src.name == newSrc.name) null "multiple" srcs; in
+      # Multiple duplicates should never be found as we deduplicate on every new element.
+      if duplicateSrc == "multiple" then throw "Something has gone terribly wrong"
```
Comment (balsoft) [inline src/opam.nix]:
```suggestion
      assert duplicateSrc != "multiple";
```

File: src/opam.nix [hunk 5]
```diff
@@ -382,6 +382,77 @@ in rec {
...
+      urlToName = url:
+        let
+          components = lib.strings.splitString "/" url;
+          filename = lib.last components;
+          name = builtins.head (lib.strings.splitString "." filename);
+        in name;
+      defToSrc = { version, ... }@pkgdef:
+        let
+          inherit (getUrl bootstrapPackages pkgdef) src;
+          name = if pkgdef ? dev-repo then urlToName pkgdef.dev-repo else "";
```
Comment (balsoft) [inline src/opam.nix]: Is having `""` as the fallback `name` for the case when there's no `dev-repo` the same as what opam-monorepo does?

**Comment by balsoft** [comment]
Overall a very neat implementation, seems to be minimal and mostly agree with upstream. Amazing job! The comments are mostly nitpicking, feel free to ignore them if you see fit. Please squash to one (or maybe a couple) commits before I merge this.

File: README.md (hunk 1 above)
Comment (RyanGibb) [reply README.md]: Yep! I wasn't sure whether to include it for the sake of completeness but on second thought I'll remove it.

File: src/opam.nix (hunk 2 above)
Comment (RyanGibb) [reply src/opam.nix]: Very good point. I'll implement this.

File: src/opam.nix (hunk 5 above)
Comment (RyanGibb) [reply src/opam.nix]: opam-monorepo filters out 'virtual' packages (without a dev-repo or url) here: https://github.com/tarides/opam-monorepo/blob/9262e7f71d749520b7e046fbd90a4732a43866e9/lib/lockfile.ml#L104

I'm still thinking about the best way of doing this with opam-nix.

This isn't the only filtering opam-monorepo does: it also filters packages that aren't build with dune (which is the reason for [opam-overlays](https://github.com/dune-universe/opam-overlays)). But AFAIK we can't do this with `opam admin list`. Maybe modifying the repo to remove any packages not built with dune is the best way to go.

I'm not sure if we need to filter packages built with dune if out of this holds true:
> Package versions from earlier repositories take precedence
over package versions from later repositories.

But if this could take a higher version from a later repository (if the higher version doesn't have an overlay yet) it would cause trouble.

**Comment by RyanGibb** [comment]
Thank you! The comments are very useful :-)

I'm actually working on a `buildOpamMonorepo` to pick up pin-depends (for example in [mirage-www](https://github.com/mirage/mirage-www/blob/62f61efe9fdacb8f260bb460691648dd34c79c9b/mirageio.opam#L42-L48)) so will convert this to a draft this and a few other tweaks are complete.

File: src/opam.nix (hunk 5 above)
Comment (balsoft) [reply src/opam.nix]: > if the higher version doesn't have an overlay yet

Hm, yes, good point. The filtering can probably be done by doing `listRepo` on the `opam-overlays` and removing those packages from `opam-repository ` entirely, before passing the final repo to `opam admin list`. This can be quite slow though.

File: src/opam.nix [hunk 6]
```diff
@@ -382,6 +384,100 @@ in rec {
...
+        let
+          inherit (getUrl bootstrapPackages pkgdef) src;
+          name = let n = nameFromURL pkgdef.dev-repo "."; in
+                 # rename dune so it doesn't clash with dune file in duniverse
+                 if n == "dune" then "_dune" else n;
+        in
+        # filter out pkgs without dev-repos
+        if pkgdef ? dev-repo then { inherit name version src; } else { };
+      # remove filterPkgs
+      filteredDefs = builtins.removeAttrs defs filterPkgs;
```
Comment (balsoft) [inline src/opam.nix]: Why is this needed here? If callers of this function want to remove some packages, it's really easy for them to just call `remoteAttrs` themselves.

File: src/opam.nix [hunk 7]
```diff
@@ -382,6 +384,100 @@ in rec {
...
+      repo = makeOpamRepo' recursive project;
+      latestVersions = mapAttrs (_: last) (listRepo repo);
+
+      pinDeps = concatLists (attrValues (mapAttrs
+        (name: version: getPinDepends repo.passthru.pkgdefs.${name}.${version})
+        latestVersions));
+    in queryToMonorepo {
+      repos = [ repo ] ++ optionals pinDepends pinDeps ++ repos;
+      # filter all query packages from monorepo
+      filterPkgs = filterPkgs ++ [ builtins.attrNames query ];
```
Comment (balsoft) [inline src/opam.nix]: `[ builtins.attrNames query ]` is a list from two elements. You probably want `[ (builtins.attrNames query) ]`

Comment (balsoft) [reply src/opam.nix]: Also, please `inherit` stuff from `builtins` or `lib` at the top of the file, and then just use the functions directly without `builtins`.

File: src/opam.nix (hunk 6 above)
Comment (RyanGibb) [reply src/opam.nix]: My thinking was this allows packages to be removed based on their name (rather than their dev repo name). For example, the unikernel package we're building has a `dummy' dev-repo which we need to remove.

File: src/opam.nix (hunk 7 above)
Comment (RyanGibb) [reply src/opam.nix]: Yep, using `(builtins.attrNames query)` fixed it! And will do.

File: src/opam.nix (hunk 6 above)
Comment (RyanGibb) [reply src/opam.nix]: This is quite useful when we want to exclude dependencies from the monorepo that are in the same source as the unikernel (e.g. https://github.com/mirage/mirage-www) in `buildOpamMonorepo` with:
>  filterPkgs = ... (attrNames (latestVersions // query)) ...;

File: src/opam.nix (hunk 5 above)
Comment (RyanGibb) [reply src/opam.nix]: I've implemented this [here](https://github.com/tweag/opam-nix/pull/18/commits/5ac390bff6a7c144ef2164739e5f5b58d0041a82), but unfortunately it doesn't consider the case when the newer version *does* build with dune but an older version (which is overlayed) doesn't.

For example, [bheap.1.0.0](https://opam.ocaml.org/packages/bheap/bheap.1.0.0/) doesn't build with dune and therefore has an overlay, but [bheapversion.2.0.0](https://opam.ocaml.org/packages/bheap/bheap.2.0.0/) does build with dune so doesn't have an overlay.

When building [mirage-hello](https://github.com/RyanGibb/mirage-hello) using `opamRepositoryModuloOverlay` we get the error
```
       > [ERROR] No solution including optional dependencies for hello.dev:   * Missing dependency:
       >             - bheap >= 2.0.0
       >             no matching version
```

I think this is why `opam-monorepo` filters (if it depends on dune) by package version rather than package name.

Comment (RyanGibb) [reply src/opam.nix]: One option I can see is parsing every package in the repo and filtering out only those that don't depend on dune. But I imagine this would be quite expensive (there are ~24k packages in `opam-repository`[0]) even if it would only have to be done once.

Another option would be to use `opam monorepo lock` to resolve dependency versions.

[0]
```
opam-repository $ find packages -mindepth 2 -maxdepth 2 | wc -l
24767
```

File: src/opam.nix [hunk 8]
```diff
@@ -480,4 +482,120 @@ in rec {
...
+        (map (pkgname: "packages/${pkgname}"))
+      ];
+    in path {
+      name = "opam-repository-modulo-overlay";
+      path = opamRepository;
+      # filter out if false, keep if true
+      filter = path: type: type != "directory" ||
+        # O(n^2)...
+        # if there isn't a match for an package in opam-overlay
+        ! elem true (map (overlayPkgPath: hasInfix overlayPkgPath path) overlayPkgPaths);
```
Comment (balsoft) [inline src/opam.nix]:
```suggestion
        ! any (overlayPkgPath: hasInfix overlayPkgPath path) overlayPkgPaths;
```

Comment (balsoft) [reply src/opam.nix]: Also, `hasInfix` might be a bit flaky here, and quite slow.

On a more general note, maybe there's a way to just make a custom `joinRepos` that would overwrite entire `package`'s instead of just package versions. That would remove the slowdown, but will probably require modifying the `symlinkJoin` from nixpkgs.

Comment (balsoft) [reply src/opam.nix]: That just made me realise that `joinRepos` is not quite correct ATM, since it doesn't actually do the version overriding properly. I think I'll fix it and implement this at the same time.

File: src/opam.nix [hunk 9]
```diff
@@ -480,4 +482,120 @@ in rec {
...
+          phases = [ "unpackPhase" "installPhase" ];
+          installPhase = ''
+            mkdir $out
+            cp -R . $out
+          '';
+        });
+    in
+    listToAttrs (map (src: nameValuePair src.name (mkSrc src)) srcs);
+
+  queryToMonorepo = { repos ? [ opamOverlays opamRepository ], resolveArgs ? { }
```
Comment (balsoft) [inline src/opam.nix]: Any reason to not use `opamRepositoryModuloOverlay` here?

File: src/opam.nix (hunk 8 above)
Comment (RyanGibb) [reply src/opam.nix]: That sounds like a good idea, but I should warn you I think it will suffer from the same issues I described in the other open comment.

Essentially there's two possibilities when there's a package in opam-overlay with a newer in opam-repository without an overlay. One, an overlay hasn't been written yet. Two, the new version now supports dune and an overlay isn't required. In the former we want to remove the new version from opam-repository, in the latter we don't.

I don't think there's a way to differentiate these cases without parsing the new version's opam file to check for a dune dependency which suggests (although I guess doesn't guarantee, but it seems to work fine for opam-monorepo) dune support.

I'm ironing out the bugs on a function to do this now. It's not fast, but I think it'll work. I'll be offline for the next couple of hours but will push what I have now.

File: src/opam.nix (hunk 9 above)
Comment (RyanGibb) [reply src/opam.nix]: Nope! Good catch

File: src/opam.nix (hunk 8 above)
Comment (RyanGibb) [reply src/opam.nix]: Okay it seems to be working! (Although it's still building... I might lose network connection soon so pushing preemptively)

Comment (balsoft) [reply src/opam.nix]: Ahh, ok, sorry for not reading thoroughly. I still made the `joinRepos` a bit better (and quite a lot faster actually) by patching `lndir` (which comes from... Xorg of all places?)

https://github.com/tweag/opam-nix/pull/20

File: src/opam.nix [hunk 10]
```diff
@@ -562,6 +562,19 @@ in rec {
 
   # remove non-dune dependant packages
   opamRepositoryFiltered =
+    # TODO is there a way to do this outside a derivation?
+    let parseOpam = pkgPath:
```
Comment (balsoft) [inline src/opam.nix]: Why not just use `importOpam`? (maybe it's a good idea to rewrite it and other functions to use your `pkgNameVer` thing, though)

File: src/opam.nix [hunk 11]
```diff
@@ -562,6 +562,19 @@ in rec {
 
   # remove non-dune dependant packages
   opamRepositoryFiltered =
+    # TODO is there a way to do this outside a derivation?
```
Comment (balsoft) [inline src/opam.nix]: No, there's not. IFD is the only way.

Comment (balsoft) [reply src/opam.nix]: (it might be a bit faster to do all the `opam2json`-ning in one derivation and then read the combined output, though).

File: src/opam.nix (hunk 10 above)
Comment (RyanGibb) [reply src/opam.nix]: Purely aesthetics. It's hard to judge progress when building  ~24k derivations with the same name 😄 I can add this to `importOpam` if you think it's nice!

File: src/opam.nix (hunk 11 above)
Comment (RyanGibb) [reply src/opam.nix]: Yeah, I was thinking the same thing. I'll give it a shot.

Comment (RyanGibb) [reply src/opam.nix]: If this works, I'm wondering if materialization might be a way of speeding up deployments...

Comment (RyanGibb) [reply src/opam.nix]: An aside: it would be nice if we could use just the version solving functionality of opam without pulling the whole program in. I might look into the feasibility of a refactor of opam to support this.

Comment (RyanGibb) [reply src/opam.nix]: Okay, I've hacked something together that builds `opamRepositoryFiltered` in one derivation:
```
  # remove non-dune dependant packages
  opamRepositoryFiltered =
    let
      keepPackages = [ "ocaml" "dune" "ocaml-solo5" "ocaml-src" "solo5"
      "ocaml-config" "ocaml-system" "ocaml-variants" "base-bytes" ];
    in bootstrapPackages.stdenv.mkDerivation {
      name = "opamRepositoryFiltered";
      src = opamRepository;
      buildInputs = with bootstrapPackages; [ jq ];
      installPhase = ''
        # look at package version directories
        # (e.g. /nix/store/<hash>-opam-repository-modulo-overlay/packages/<name>/<name>.<version>)
        for f in $(find packages -mindepth 2 -maxdepth 2); do
          echo $f
          # special cases as we will filter them before creating monorepo
          # just needed so `opam admin list` doesn't complain
          # TODO parse queries opam files to a list of dependancies with 
          # build=false so only pick up monorepo deps and we remove this workarounds
          # TODO or just use opam monorepo solver...
          if $(echo "$f" | awk 'BEGIN {
              split("${concatStringsSep " " keepPackages}", parts);
              for (i in parts) dict[parts[i]]=""
            } { exit !($1 in dict) }'); then
            # do nothing
            echo
          # if package depends on dune
          # TODO properly recursively parse vals of depends
          elif $(${opam2json}/bin/opam2json "$f/opam" |\
            jq '[ .depends | flatten | .[] | if type=="string" then . else .val end ] |
              if index("dune") == null then halt_error(1) else halt end'); then
              mkdir -p "$out/$f"
              cp -r "$f" "$out/$f"
          fi
        done
      '';
    };
 ```
 But it's also *very* slow (waiting on exact timings now). I think using `opam monorepo`'s solver would be a better solution.

Comment (RyanGibb) [reply src/opam.nix]: Okay, this creates a very IO bound process that takes ~half an hour to complete 😄 I'm going to look at using `opam monorepo lock` and parsing the lockfile to solver versions.

Comment (balsoft) [reply src/opam.nix]: That sounds reasonable.

**Comment by RyanGibb** [comment]
So trying use `opam monorepo lock` has more problems than I anticipated. First and foremost I can't get it to [use local repositories](https://github.com/tarides/opam-monorepo/issues/333).

However I don't think we actually *need* this. It's nice to have, but a work around for:
> The case when the newer version _does_ build with dune but an older version (which is overlayed) doesn't.

Is just to downgrade opam-repository.

This isn't ideal, but I'm happy to merge this as-is if you are and create a separate PR for this customer solver logic. Alternatively, I can carry on using the fork and let you know when I have more success.

**Comment by RyanGibb** [comment]
An aside: another use case I found for `opam monorepo` is picking up [`or` dependencies](https://github.com/RyanGibb/mirage-www/issues/1#issuecomment-1254916538). A workaround for this is just manually adding them to the query.

**Comment by balsoft** [comment]
> This isn't ideal, but I'm happy to merge this as-is if you are and create a separate PR for this customer solver logic

Yep, I think this is a good-enough solution as it is right now. `opam monorepo lock` can be added in a later PR.

**Comment by RyanGibb** [comment]
Great :-)

**Comment by balsoft** [comment]
Could you clean up the commit history then?

**Comment by RyanGibb** [comment]
Done!

File: README.md [hunk 12]
```diff
@@ -582,6 +582,63 @@ Then, import it:
...
+<div class=example id=build-opam-project-all dir=my-package>
+
+
+
+```nix
+(buildOpamProject' { } ./. { }).my-package
+```
+
+</div>
+
```
Comment (balsoft) [inline README.md]: This whole example is incorrect. It needs a different `id`, also most likely a different `dir` and also the Nix code of the example is wrong.

Comment (balsoft) [reply README.md]: (the new folder should be in `examples/readme`, and should contain the opam-monorepo opam file and so on)

Comment (RyanGibb) [reply README.md]: Whoops, thanks for pointing that out.

Given that this is a relatively specialist use case that is most likely going to be used from https://github.com/RyanGibb/hillingar I'm happy to leave this without an example project if you are.

File: README.md [hunk 13]
```diff
@@ -582,6 +582,49 @@ Then, import it:
...
+Similar to `queryToScope`, but creates a attribute set (instead of a
+scope) with package names mapping to sources for replicating the
+[`opam monorepo`](https://github.com/tarides/opam-monorepo) workflow.
+
+The `filterPkgs` argument gives a list of package names to filter from
+the resulting attribute set, rather than removing them based on their
+opam `dev-repo` name.
+
+#### `buildOpamMonorepo'`
```
Comment (balsoft) [inline README.md]: AFAIU the function is `buildOpamMonorepo`

File: README.md [hunk 14]
```diff
@@ -582,6 +582,49 @@ Then, import it:
...
+
+```
+{ repos = ?[Repository]
+; resolveArgs = ?ResolveArgs
+; pinDepends = ?Bool
+; recursive = ?Bool
+; extraFilterPkgs ?[ ] }
+→ project: Path
+→ Query
+→ Scope
```
Comment (balsoft) [inline README.md]: This doesn't quite return a `Scope`, it returns an attrset of sources, doesn't it?

File: README.md [hunk 15]
```diff
@@ -582,6 +582,49 @@ Then, import it:
...
+```
+
+A convenience wrapper around `queryToMonorepo`, similar to
+`buildOpamMonorepo'`.
```
Comment (balsoft) [inline README.md]: ???

File: README.md (hunk 12 above)
Comment (balsoft) [reply README.md]: Hm, ok. Would you consider moving the functionality of `hillingar` to `opam-nix` in a future PR? I'm happy to give you commit access to opam-nix to simplify maintenance.

File: README.md (hunk 15 above)
Comment (RyanGibb) [reply README.md]: Yeah, apologies, I think I intended to come back and proof read these but it slipped my mind 😅

File: README.md (hunk 12 above)
Comment (RyanGibb) [reply README.md]: Yep, I would, that sounds like it would be sensible. I created it as a separate project as it's quite mirage specific (and I wasn't sure if there would be an appetite to incorporate it into opam-nix) but also it's still quite volatile as I'm patching it to work for different unikernels. After it's more stable I'll create a separate PR for it.
